src/Microsoft.OData.Client/Serialization/Serializer.cs

```diff
@@ -358,12 +358,162 @@ internal void WriteEntry(EntityDescriptor entityDescriptor, IEnumerable<LinkDesc
+                {
+                    relatedEnd = item as LinkDescriptor;
+                }
+
+                navigationLink.Name = relatedEnd.SourceProperty;
+                bool isCollection = clientType.GetProperty(relatedEnd.SourceProperty, UndeclaredPropertyBehavior.ThrowException).IsEntityCollection;
+
+                odataWriter.WriteNestedResourceInfoStart(navigationLink);
+
+                odataWriter.WriteNestedResourceInfoStart(navigationLink, relatedEnd.Source, relatedEnd.Target);
+
+                if (isCollection)
+                {
+                    WriteResourceSet(relatedLinks.Value, linkDescriptors, bulkUpdateGraph, odataWriter);
+                }
+                else
+                {
+                    //we can have a related single entry ~ this will automatically be a replace operation.
+                    if (item.DescriptorKind == DescriptorKind.Entity)
+                    {
+                        EntityDescriptor relatedEntityDescriptor = item as EntityDescriptor;
+                        WriteDeepInsertEntry(relatedEntityDescriptor, linkDescriptors, bulkUpdateGraph, odataWriter);
+                    }
+                    else if (item.DescriptorKind == DescriptorKind.Link)
+                    {
+                        odataWriter.WriteEntityReferenceLink(new ODataEntityReferenceLink() { Url = navigationLink.Url }, relatedEnd.Source, relatedEnd.Target);
+                    }
+                }
+
+                odataWriter.WriteNestedResourceInfoEnd(navigationLink, relatedEnd.Source, relatedEnd.Target);
+                odataWriter.WriteNestedResourceInfoEnd();
+            }
+        }
+
         /// <summary>
         /// Writes the feed entry
         /// </summary>
-        /// <param name="descriptors">A list of descriptors to be written</param>
-        /// <param name="linkDescriptors">Descriptor links for some descriptors.</param>
-        /// /// <param name="bulkUpdateGraph">An instance of the <see cref="BulkUpdateGraph"/></param>
+        /// <param name="descriptors">A list of descriptors to be written.</param>
+        /// <param name="bulkUpdateGraph">An instance of the <see cref="BulkUpdateGraph"/>.</param>
+        /// <param name="oDataWriter">An instance of the <see cref="ODataWriterWrapper"/>.</param>
+        internal void WriteResourceSet(IReadOnlyList<Descriptor> descriptors, Dictionary<Descriptor, List<LinkDescriptor>> linkDescriptors, BulkUpdateGraph bulkUpdateGraph, ODataWriterWrapper oDataWriter)
+        {
+            ODataResourceSet resourceSet = new ODataResourceSet();
+            oDataWriter.WriteStart(resourceSet);
+
+            for (int i = 0; i < descriptors.Count; ++i)
+            {
+                Descriptor descriptor = descriptors[i];
+
+                if (descriptor is EntityDescriptor entityDescriptor)
+                {
+                    if (entityDescriptor.State == EntityStates.Deleted || entityDescriptor.State == EntityStates.Modified)

```

Comment:
When building the graph, didn't we already check for this and throw exceptions where applicable? Do we need to perform the same checks here?